nixos/modules/virtualisation/oci-containers.nix

```diff
@@ -247,192 +254,311 @@ let
 
         preRunExtraOptions = mkOption {
           type = with types; listOf str;
-          default = [];
+          default = [ ];
           description = "Extra options for {command}`${defaultBackend}` that go before the `run` argument.";
-          example = [ "--runtime" "runsc" ];
+          example = [
+            "--runtime"
+            "runsc"
+          ];
         };
 
         extraOptions = mkOption {
           type = with types; listOf str;
-          default = [];
+          default = [ ];
           description = "Extra options for {command}`${defaultBackend} run`.";
           example = literalExpression ''
             ["--network=host"]
           '';
         };
 
         autoStart = mkOption {
-          type = types.bool;
+          type = with types; bool;
           default = true;
           description = ''
             When enabled, the container is automatically started on boot.
             If this option is set to false, the container has to be started on-demand via its service.
           '';
         };
+
+        pull = mkOption {
+          type =
+            with types;
+            enum [
+              "always"
+              "missing"
+              "never"
+              "newer"
+            ];
+          default = "missing";
+          description = ''
+            Image pull policy for the container. Must be one of: always, missing, never, newer
+          '';
+        };
+
+        capAdd = mkOption {
+          type = with types; lazyAttrsOf (nullOr bool);
+          default = { };
+          description = ''
+            Capabilities to add to container
+          '';
+          example = literalExpression ''
+            {
+              SYS_ADMIN = true;
+            {
+          '';
+        };
+
+        capDrop = mkOption {
+          type = with types; lazyAttrsOf (nullOr bool);
+          default = { };
+          description = ''
+            Capabilities to drop from container
+          '';
+          example = literalExpression ''
+            {
+              SYS_ADMIN = true;
+            {
+          '';
+        };
```

Comment:
Couldn't these be merged into a single `capabilities` option?
- `true` -> add
- `false` -> remove
- `null` or omitted -> default

Reply:
That's technically possible, but in my humble opinion it would be a more confusing interface.  Part of this module's appeal is that it is a fairly thin abstraction over the underlying container runner implementations; podman and docker (and their various documentation and examples) have separate `--cap-add` and `--cap-drop` options so it seems reasonable for us to do the same.

Reply:
Various declarative wrappers over Docker/podman (Ansible, docker-compose etc) also use separate cap_add and cap_drop options

Reply:
A practical benefit of modeling it that way is that you can override the capabilities using the module system.

It is also conceptually simpler, because you don't have to make users bend to the quirks of an `listOf str`-typed CLI. That's just not a good data model. I shouldn't be able to express this, because it has no clear meaning:

```
capAdd.SYS_ADMIN = false;
capDrop.SYS_ADMIN = false;
```

> Various declarative wrappers over Docker/podman (Ansible, docker-compose etc) also use separate cap_add and cap_drop options

Arion - a declarative docker-compose wrapper - does not, and I've given that as an example.
I haven't heard of another wrapper; would like to know more.

Apologies if I come across a bit sour. GitHub marked my earlier comment as outdated, so I'll blame them for cutting the [conversation](https://github.com/NixOS/nixpkgs/pull/349537/files#diff-356b377080b6689db2cc419c736399adb0725b0a2c98da1900b30085e877ccb7) short :)
I'd be happy to propose a refactor.

Reply:
Something like this?
https://github.com/NixOS/nixpkgs/pull/363574

Reply:
You know what, after reading #363574 I think I actually withdraw my objection.  That interface and option description make it quite understandable.